### ISSUE TITLE
api: improve API for bucket creation

### ISSUE THREAD

**Issue by nharraud** [inveniosoftware/invenio-records-files#21]
**Problem:**
It is not intuitive that accessing `FilesMixin.files` will create a bucket if there is no bucket. A getter should not modify the database.

For B2Share deposits we want to create a bucket for every deposit. The bucket can later be deleted if the deposit is published without any file. The only way to do that with FilesMixin API is to call `deposit.files`. This looks more like a dead line of code than the creation of a bucket.

What's more, there is no way to know if a bucket exist with the API. One has to directly query the model. If, like I did, somebody tries to see if there are files in the serializer of invenio-records-rest, a bucket will be created automatically instead of saying that there are no files. This can create important bugs in case of rollback.

**Suggestion:**
- Add a method for creating the bucket. It could be a setter which takes a dict like this:

```
record.files = { 'hello.txt': BytesIO(b'Hello world!') }
```
- The `files` getter should not create a bucket if there is none. It can either return an iterator which ends immediately or return None. If an iterator is returned the bucket of the iterator should be `None`. That way it is explicit that there are no bucket.

**Comment by nharraud** [comment]
API design issue
cc @inveniosoftware/triagers

**Comment by nharraud** [comment]
related to inveniosoftware/invenio-records-files#23

**Comment by lnielsen** [comment]
FYI: On Zenodo I've implemented that my custom deposit class creates the bucket. This is needed in order to make the bucket available immediately via the Files-REST API. Also, if a deposit is deleted without being published, the bucket will be completely removed.

**Comment by nharraud** [comment]
@lnielsen This is also [what we are currently doing in B2Share](https://github.com/EUDAT-B2SHARE/b2share/blob/33e49cdc6036e608acde18b78df848ce41f802dd/b2share/modules/deposit/api.py#L107). But the problem is not that it is not possible, it is that we have IMHO a confusing and inefficient behavior by lazily creating the bucket when `record.files` is called.

*Example of performance issues:*
in `invenio-deposit` [we call `deposit.files`](https://github.com/inveniosoftware/invenio-deposit/blob/352336ba2b7786369b445fb3c4b34669e2ffad3a/invenio_deposit/api.py#L249) to check if the deposit has files. This will create a bucket which will be discarded right away if there are no files (see [here](https://github.com/inveniosoftware/invenio-records-files/blob/c46676d91ed66444981fb44d16849d51204d347f/invenio_records_files/api.py#L243) and [here](https://github.com/inveniosoftware/invenio-deposit/blob/352336ba2b7786369b445fb3c4b34669e2ffad3a/invenio_deposit/api.py#L502)). Thus we have multiple calls to the DB without any need as [we already knew here](https://github.com/inveniosoftware/invenio-records-files/blob/c46676d91ed66444981fb44d16849d51204d347f/invenio_records_files/api.py#L239) that the deposit has no files.

*Example of confusing and error prone behavior:*
Record serializers are called [outside of the main transaction](https://github.com/inveniosoftware/invenio-records-rest/blob/master/invenio_records_rest/views.py#L507) as they are not supposed to change the data. Thus they can't do something like the following example because the bucket id would be discarded as soon as the response is sent:
```python
if record.files:
  links['file_upload_link'] = url_for(
    'invenio_files_rest.bucket_api',
    bucket_id=record.files.bucket
  )
```
And even if the bucket id was not used, calling `record.files` to check files existence would still call the DB at each serialization.

We could say that `record.files` should not be used to just check if there are any files, but as seen above we already do it in our code. It's confusing if sometime it is ok to use it that way and other times it is not a good idea.

Note that it is possible to check the bucket existence by using `invenio-records-files` models directly. One simple solution which I would IMHO find better is to always use the `Bucket` and `RecordsBuckets` models to create the buckets, and use `invenio-records-files` only for *reading* the buckets or checking that they exist, but not for creating them.

**Comment by lnielsen** [comment]
@nharraud My comment was not to say that the current module is fine as-is :-) In fact, rather that I had the same need :-)

I do think it makes sense for Invenio-Deposit to create the bucket up-front (or not if you don't need file upload).

**Comment by nharraud** [comment]
Summary of a discussion with @jirikuncar:
As `record.files` creates a bucket by calling `record._create_bucket`, and its default implementation does not create any bucket, we can just add a warning to the documentation of `record._create_bucket` or `record.files` explaining that if `record._create_bucket` is overriden and creates a bucket, `record.files` should **NEVER** be used to check if the record has files.

**Comment by lnielsen** [comment]
> ``record.files`` should NEVER be used to check if the record has files

That doesn't seem intuitive to me. Especially since you can do things like e.g. ``len(record.files)``, then it seems natural to do e.g. ``if record.files``.

**Comment by jirikuncar** [comment]
> `record.files` should **NEVER** be used to check if the record has files.

I would like to clarify this since I was mentioned here. I agree that side effect, while checking the property, is not good. However it is there only when `Record._create_bucket` is implemented, hence it's kind of expected and my recommendation was: "Do not to use `files` property if you have implemented `_create_bucket` method and you don't want bucket to be created".

Anyway we will have to re-think the API for "multi bucket" support in the future but for now I am in favor of keeping the API as-is and wait for more feedback.

**Comment by fenekku** [comment]
I know this thread is old / closed, but wanted to +1 the desire for a no-side-effect `FilesMixin.files` getter. 

 I ran into this when trying to index `'_files'` metadata through an [index receiver (hook)](https://github.com/inveniosoftware/training/tree/master/10-indexing-records) when creating a deposit. At that time, the `RecordsBuckets` entry is not created yet (it will be *just* after), so when calling 

```python
bucket = Bucket.get(json.get('_buckets', {}).get('deposit'))
json['_files'] = deposit.files.dumps(bucket=bucket)  # this creates a bucket
```
I get a bucket created out of the blue, that will not be used. 

Obviously, the problem is a little more systematic: 
- invenio_deposit defines `_create_bucket`
- `deposit.files.dumps(bucket=bucket)` is how to dump files metadata
- a hook is used for this

**Comment by lnielsen** [comment]
Closed by #83 

- Old behavior of creating the bucket when accessing files has been removed.
- PR #83 breaks Invenio-Deposit.
- The Record class now by default creates a bucket when creating the record.
